### ISSUE TITLE
Input device (or joystick/game controller) portal

### ISSUE THREAD

**Issue by smcv** [flatpak/xdg-desktop-portal#536]
This is semi-redundant with #227, but its narrower scope might make it more achievable than a fully general USB device portal.

For brevity I'm going to use "joystick" the same way the kernel and udev do, to refer to any gaming input device (gamepad, flight stick, steering wheel, whatever).

## Use cases

[GNOME Games](https://github.com/flathub/org.gnome.Games) wants to enumerate joysticks' evdev device nodes and use them to control games, via [libmanette](https://gitlab.gnome.org/aplazas/libmanette/).

Many games want to enumerate joysticks' evdev device nodes and use them to control games, often via libraries like SDL ([SDL joystick enumeration](https://hg.libsdl.org/SDL/file/tip/src/joystick/linux/SDL_sysjoystick.c), [SDL udev wrappers](https://hg.libsdl.org/SDL/file/tip/src/core/linux/SDL_udev.c)).

Steam wants to talk to raw HID devices that correspond to joysticks to perform input device remapping.

Wine/Proton (when run by Steam, or by Lutris or similar) wants to use raw HID devices for joysticks in preference to evdev device nodes, under at least some circumstances and for at least some joysticks, because some Windows games bypass higher layers and talk directly to the HID device to get finer control over Xbox and Playstation controllers. ([Upstream Wine](https://github.com/wine-mirror/wine/tree/master/dlls/winebus.sys), [Proton's derived version](https://github.com/ValveSoftware/wine/blob/proton_5.0-next/dlls/winebus.sys/bus_udev.c))

Wine/Proton has relatively elaborate heuristics to get the most appropriate/compatible vendor IDs, product IDs, revision numbers etc. for game controllers that were detected via libudev, so it needs access to various libudev device properties of the device and its HID/USB/input ancestor devices as metadata - knowing the existence or nonexistence of a device is not enough.

## Current practice

The Flatpak apps for Steam and GNOME Games both run with `--device=all` to share the entire host system `/dev`. Ideally they would not have to do this.

libmanette, SDL and Wine/Proton use `libudev` to enumerate and monitor input devices. However, this is not designed to be usable inside a container, with two main problems:

* the protocol between `udevd` and `libudev` is not designed to be backwards- or forwards-compatible, so it's a matter of luck whether the host `udevd` is sufficiently close to be compatible with the container `libudev`
* netlink events are authenticated by checking their uid is 0, but a non-setuid bubblewrap is unable to share uid 0 with the container's userns, so the netlink events necessary for detecting controller hotplug are not available

libmanette can be compiled in a fallback mode where it watches `/dev/input` with inotify. However, this only gets information from the kernel, not the synthetic properties added in user-space by udevd (in particular `ID_INPUT_JOYSTICK` to distinguish between joysticks and other devices). libmanette assumes that any evdev input device that it can open read/write is a game controller, which is *usually* true, but will be false if the user has added themselves to the `input` group that owns evdev devices, resulting in unrelated input devices being misidentified as game controllers.

SDL can be compiled in a fallback mode where it watches `/dev/input` by polling periodically. However, this only gets information from the kernel, as above. It can also fail to detect if `/dev/input/event23` is unplugged, a new device is plugged in, and the new device happens to be assigned the same device node as the old device, `/dev/input/event23`. SDL *does* have heuristics to guess whether an input device is in fact a joystick, but those heuristics are not particularly clever: anything with buttons and an absolute X and Y axis is treated as a joystick, which can result in non-joystick devices like touchpads being guessed to be joysticks.

It's also not ideal for the heuristics for "what is a joystick?", and for device identification, to be duplicated in SDL, Wine and various other middleware libraries: it would be better if they could all be fed known-good information from udev.

I'm currently working on putting Steam games in containers (pressure-vessel), which reuses a lot of Flatpak code, but is not intended to be a security boundary, and I've run into many of the same issues there; so I am probably going to need to design a mechanism to get input devices from the outside world into the container *anyway*, which could serve as a prototype for this. However, I can't put a lot of time into designing permissions control/UI prompting for it at this stage, because the part of the problem I am focusing on solving is the mechanics of getting device monitoring and metadata into the container at all.

**Comment by smcv** [comment]
> talk to raw HID devices

This isn't normally allowed for unprivileged users, but Steam expects/requires that [its udev rules](https://github.com/ValveSoftware/steam-devices/) have been installed on the host system, to allow raw HID access to various known game controllers and VR devices. I know udev rules on the host are way outside the scope of what Flatpak or xdg-desktop-portal can do, but having an input device portal of the form "*if* you have opened up access on the host, then it's passed through to the container" makes this at least solvable.

Prior art in Snap: https://snapcraft.io/docs/hidraw-interface

**Comment by smcv** [comment]
Prior art in Windows UWP: Windows prevents access to keyboards, mice etc., but allows access to all other HID devices for apps whose manifest mentions `<DeviceCapability Name="humanInterfaceDevice"/>`. In versions older than Windows 10 v1809, the app had to specify which vendor/product it wanted to access; in newer versions it's all-or-nothing. https://docs.microsoft.com/en-us/uwp/api/windows.devices.humaninterfacedevice?view=winrt-19041

**Comment by hadess** [comment]
This was rejected in favour of a Wayland protocol extension.

See "inputfd":
https://gitlab.gnome.org/GNOME/mutter/-/tree/wip/carlosg/inputfd
https://lists.freedesktop.org/archives/wayland-devel/2017-August/034828.html
(which somebody needs to finish)

**Comment by smcv** [comment]
Steam supports non-Wayland systems, so I am probably going to have to implement something that looks very much like what I proposed here one way or another, regardless of whether it's desired in xdg-desktop-portal. I'll try to make it "the right shape" for an xdg-desktop-portal interface, so that it can be copied into xdg-desktop-portal if people change their minds, but I won't push to get it included in x-d-p.

Thanks for the references to inputfd - I also have to put together a library API for this, so I'll try to make sure it would be implementable in terms of the inputfd protocol on systems that happen to have it.

**Comment by Vanadiae** [comment]
> libmanette assumes that any evdev input device that it can open read/write is a game controller, which is usually true

No, libmanette does a bit of check (more less the same thing as udev does for ID_INPUT_JOYSTICK). See is_game_controller function in libmanette-monitor.c.

> Thanks for the references to inputfd - I also have to put together a library API for this, so I'll try to make sure it would be implementable in terms of the inputfd protocol on systems that happen to have it.

the inputfd protocol isn't finished (i.e. published on the wayland-protocols repository), nor is it implemented by any wayland compositor (logic) so you can't really make a dbus wrapper library for it for now.

**Comment by smcv** [comment]
> libmanette does a bit of check (more less the same thing as udev does for ID_INPUT_JOYSTICK). See is_game_controller function in libmanette-monitor.c.

Thanks, I'll take a look!

> you can't really make a dbus wrapper library for it for now

No, but I can at least make it close enough to "the same shape" that it would be *implementable*, assuming the design changes made before it gets implemented are not too drastic.

**Comment by smcv** [comment]
Ah, sorry, I was looking in the wrong place - I assumed libmanette's filtering would be in the Monitor, but it's actually in the Device.

**Comment by Vanadiae** [comment]
> Ah, sorry, I was looking in the wrong place - I assumed libmanette's filtering would be in the Monitor, but it's actually in the Device.

Huh, that's not normal, it should live in the monitor (the same way as udev's filtering is done there). If a device is marked manually as ID_INPUT_JOYSTICK in udev's hwdb because it doesn't match udev's automatic tagging, then it won't be accepted as game controller because it won't match the check in libmanette too since it's basically the same as udev's automatic tagging. I'll open a bug for this and address it later on in a MR.

Thanks for noticing this.

**Comment by swick** [comment]
Just going to mention that it might be a better long term plan to instead adopt an OpenXR input action style API.

**Comment by hadess** [comment]
> Just going to mention that it might be a better long term plan to instead adopt an OpenXR input action style API.

What's an "OpenXR input action style API"?

**Comment by hadess** [comment]
After discussing this @smcv on IRC, I think that it should be possible to both have a Wayland protocol, as well as a simpler enumeration/opening of event nodes over D-Bus, as he's [already prototyped](https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/merge_requests/158).

Having a D-Bus based API would allow to have a fallback for non-Wayland clients, Wayland compositors without the support for that protocol, or simply non-Wayland compositors (X11).

There are a number of required features for this to work out:
- a simple D-Bus API that wouldn't require much work to call from raw C, I'm thinking of SDL here. It might very well be easier to create a shim udev library that hides the enumeration side of things, though opening the device node would still need to go through another API. I'd recommend using sd-bus to avoid "unwanted" threads.
- it needs revocation (`EVIOCREVOKE`) support, so that it's possible to yank the joypad fd from one client to give it to another, and not have 2 apps using/sniffing data from the joypad at the same time
- the metadata available outside the sandbox needs to be made available without requiring new API to exist. For example, if the kernel suddenly started tagging PlayStation controllers with `JOYPAD_LAYOUT=sony` so that X/triangle/square/circle is used in popups, we would need this to be available to libraries and apps straight away
- a clear idea on how to handle rumble, and LEDs, as well as composite devices, like the Dualshock4 device with its touchpad, and accelerometer.

**Comment by smcv** [comment]
https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/merge_requests/159 is a better link for my portal prototype, !158 is just the API. The way the portal works is that the server-side uses an implementation of the SrtInputDeviceMonitor GInterface from !158 to watch the actual devices, and the client-side also presents a (different concrete implementation of) SrtInputDeviceMonitor.

> a simple D-Bus API that wouldn't require much work to call from raw C

There's a trade-off between this on one hand, and making the API follow portal API conventions and allow for extensibility on the other. The setup with Request and Session objects took me a while to understand and get right, although I can see why it's desirable; it would be good to assess whether the Request is really needed, when dbus-daemon (and presumably dbus-broker?) is in fact willing to put an arbitrarily large timeout on calls.

Many of the sort of libraries that would want to avoid GDBus are unfortunately also the sort of libraries that don't have a particularly well-defined event loop abstraction like GLib does, which is going to make async calls painful, and signals more so - but we really need to be using either async calls or signals if we want to be prompting the user for permission, and in any case we need signals for monitoring. Doing GDBus (or maybe sd-bus) in a private GMainContext on a non-main thread might actually be *easier* than trying to stuff this into the main thread.

I personally don't actually need to deal with prompting right now, because my use-case has `--devices=all` or the equivalent anyway, but I recognise that we ought to accommodate things that are more-meaningfully-sandboxed as well.

> the metadata available outside the sandbox needs to be made available without requiring new API to exist

There are two sides to this, which we should keep aware of. One is the metadata that genuinely comes from the kernel: we can often read this from `/sys` or from the input fd itself, and don't *necessarily* need to proxy it all across D-Bus or Wayland. (But we have it available to us, so we might as well.)

The other is the metadata that comes from user-space (udev), such as the `ID_INPUT_FOO` properties, which container processes cannot get any other way, and *must* get from the portal if we want to provide that information.

> it needs revocation (EVIOCREVOKE) support

The only way my current prototype could represent this would be to revoke the fd returned by OpenDevice(), send the Removed signal for the device, and generally behave in a way that the game cannot distinguish from the joystick being physically unplugged.

I'm not sure how good that would be for local (split-screen) multiplayer in particular: if Alice and Bob are both using indistinguishable Xbox controllers (assume they're cheap clone controllers with no meaningful serial number), and they change compositor focus (causing input fd revocation) then change back to the game, we don't want to send a pair of Added signals in arbitrary order that end up with Alice controlling Bob's character and vice versa :-)

So maybe we need a representation for controllers that sort-of-exist, but are currently inaccessible?

> a clear idea on how to handle rumble, and LEDs

If we pass through the evdev fd, the only way we can handle this is "the game has total control", because I'm fairly sure you do both by poking the fd with ioctls.

If that's unacceptable, then I think the only way would be for the game to be given something that is more like an AF_UNIX socket, a uinput-synthesized input device (normally requires privileges, although Steam currently relies on this anyway), or a CUSE device (requires privileges, I think). In the case of an AF_UNIX socket, the game would also have to know how to speak a portal-specific protocol over it (perhaps heavily based on evdev), instead of using evdev in the usual way.

> as well as composite devices, like the Dualshock4 device with its touchpad, and accelerometer

The touchpad comes out to user-space as a pointing device (it's effectively a laptop touchpad) so I don't think we (as in unprivileged user-space in general) get to have direct access. Games will just have to cope with getting their mouse events through X11 and Wayland, like they do for USB and PS/2 mice.

The accelerometer is a separate fd, which SrtInputDevice will try to tag with SRT_INPUT_DEVICE_TYPE_FLAGS_ACCELEROMETER (our equivalent of udev's ID_INPUT_ACCELEROMETER), either based on udev tags or guessing from the evdev capabilities. If games want to correlate the accelerometer with its matching gamepad device (so that two players can each use a combination of gamepad and motion controls), they can do that by looking at the udev metadata that SrtInputDevice exposes - I think matching up the `input_ancestor_sys_path` is probably the right thing. Or, if they can rely on having udev and libinput on the host, they could get the libinput device groups from the udev properties. Either way, they'd have to do the same thing if they used the input device nodes directly.

**Comment by smcv** [comment]
> Just going to mention that it might be a better long term plan to instead adopt an OpenXR input action style API

This is an interface in which the game receives higher-level, already-remapped input events from a user-space component, instead of directly accessing evdev devices provided by the kernel?

For games that use evdev, I'm not sure how popular that would be: using evdev is well-understood, even if not straightforward, and is a broadly-applicable de facto standard. Steam's Steam Controller and Remote Play support use uinput to synthesize fake input devices, rather than relying on the game using particular middleware that understands how to get input events from a user-space component. I suspect that's sufficiently painful to achieve that the developers responsible wouldn't have done it if they saw a good alternative.

For Wine/Proton specifically, Proton tries hard to use raw HID (`hidraw`) devices to access input devices directly, bypassing evdev. This is because one of the Windows APIs that it is emulating involves games opening HID devices directly, for example if they want low-level access to Playstation controllers. It does *have* code to open a SDL or evdev device and synthesize plausible HID responses based on what it sees, but my understanding is that this is not exactly reliable or seamless. So it's desirable to be able to make raw HID devices available, at least for apps with appropriate permissions set.

**Comment by swick** [comment]
> This is an interface in which the game receives higher-level, already-remapped input events from a user-space component, instead of directly accessing evdev devices provided by the kernel?

Yes.

> For games that use evdev, I'm not sure how popular that would be: using evdev is well-understood, even if not straightforward, and is a broadly-applicable de facto standard. 

Evdev has been the only way to access controllers so obviously it's a de facto standard right now.

What applications don't rely on SDL or similar libraries but use evdev directly? If we patched SDL etc to use a higher level input what would be the fallout?

Either way, for the remaining cases emulating evdev seems like a solvable problem.

> Steam's Steam Controller and Remote Play support use uinput to synthesize fake input devices, rather than relying on the game using particular middleware that understands how to get input events from a user-space component.

The uinput events would end up in the component generating the OpenXR style input events. Maybe I don't get what you're implying here.

(side note: uinput for applications is horrible and should eventually be replaced by a solutions like [libei](https://who-t.blogspot.com/2020/08/libei-library-to-support-emulated-input.html)).

> For Wine/Proton specifically, Proton tries hard to use raw HID (hidraw) devices to access input devices directly, bypassing evdev. This is because one of the Windows APIs that it is emulating involves games opening HID devices directly, for example if they want low-level access to Playstation controllers.

This is not forwards compatible. Sure, it works fine if the game supports a device that the user has but in 5, 10 or 15 years users will own different input hardware and now suddenly the game won't work with that any more.

> It does have code to open a SDL or evdev device and synthesize plausible HID responses based on what it sees, but my understanding is that this is not exactly reliable or seamless. So it's desirable to be able to make raw HID devices available, at least for apps with appropriate permissions set.

Maybe the emulation should be improved instead of investing and cementing a system which is obviously not forwards compatible?

The OpenXR style API might even help a lot with the emulation: if you want to emulate a specific device you can ask the API for the correct set of actions.

**Comment by hadess** [comment]
> There's a trade-off between this on one hand, and making the API follow portal API conventions and allow for extensibility on the other. The setup with Request and Session objects took me a while to understand and get right, although I can see why it's desirable; it would be good to assess whether the Request is really needed, when dbus-daemon (and presumably dbus-broker?) is in fact willing to put an arbitrarily large timeout on calls.

I don't know, you'd need to ask @matthiasclasen about this I'm afraid.

> I'm not sure how good that would be for local (split-screen) multiplayer in particular: if Alice and Bob are both using indistinguishable Xbox controllers (assume they're cheap clone controllers with no meaningful serial number), and they change compositor focus (causing input fd revocation) then change back to the game, we don't want to send a pair of Added signals in arbitrary order that end up with Alice controlling Bob's character and vice versa :-)

The device node, outside the sandbox stays the same, so the portal should be able to present them in the same order than it did before. It could even generate a fake identifier if it sees duplicate devices.

> > a clear idea on how to handle rumble, and LEDs
> 
> If we pass through the evdev fd, the only way we can handle this is "the game has total control", because I'm fairly sure you do both by poking the fd with ioctls.

LEDs use a different API (they go through sysfs, not device nodes). Unless you want to write all the drivers in user-space that is, as Steam is wont to do for a number of devices. It's usually not a big deal in most cases for LEDs not to be handled, as the kernel does something useful with them by default (colours on PlayStation 4 controllers, player numbers on PS3 and Wii).

**Comment by HugLifeTiZ** [comment]
Hey there. I feel like I may be barging into a conversation that I'm not really sure I belong in, since I'm more of an end user that occasionally contributes whenever it's in my ability, but I've recently regained interest in continuing to pursue the problem of where there are sporadically certain flatpak apps and Steam games within the Steam flatpak that can't read game controllers, documented here:

* [Issue in Steam flatpak github](https://github.com/flathub/com.valvesoftware.Steam/issues/498)
* [Thread on Flathub forum](https://discourse.flathub.org/t/gamepads-not-working-in-various-applications/518/2)
* [Issue on Firefox's tracker](https://bugzilla.mozilla.org/show_bug.cgi?id=1645955) that I just kind of let stagnate like a bad reporter :cry:

To clarify what's going on, it seems that apps and games that use SDL2 can see my gamepads just fine. But if they use SDL1 or some other means of reading the game controllers, they can't see gamepads at all. They can't even see the virtual gamepads that Steam Input makes. So you've got this weird disconnect where Steam itself is happily reading and processing inputs, but another program in the same sandbox is entirely oblivious to them. My system disables Steam Input (rather, prevents it from looking at real controllers) in favor of MoltenGamepad, but it seems that regardless of whether the gamepads are filtered via Steam Input, filtered via MoltenGamepad, or left completely raw, some programs just can't see them *at all*.

I've recently found that gamepads can't be read in Ungoogled Chromium's flatpak, nor within browser-based overlays loaded within OBS's flatpak, be it via the external browser plugin or the internal one compiled in via modifying the manifest. I've had a particular motivation to be able to see my inputs in my recorded gameplay footage, and I'd love to use the Steam flatpak for all my Steam games, and this has become a blocker for me on both things. There is an input overlay plugin I could compile for OBS, but it's proven to be a massive pain, and for all I know, the way it's doing things could fail just the same as other apps and games have already been doing.

I could reopen the Firefox issue to hopefully provide more information and get them to look at it for more concrete answers, but I feel like coming here to discuss it is the more appropriate action, especially since some of you might already have insight on what's going on here.

I know the discussion here is primarily regarding the "right way" forward, and that's great, but do some of the issues discussed happen to overlap with what may be going on with these apps not receiving input, and is there anything I can do as an end-user/baby contributor to work toward fixing this in a more immediate term?

**Comment by smcv** [comment]
> if they use SDL1 or some other means of reading the game controllers, they can't see gamepads at all

I think this is most likely because SDL1 and those other means of reading the game controllers are relying on `udev_monitor` in `libudev`, which does not usually work in containers, because the events that it delivers do not cross the sandbox/container boundary ([more info here](https://bugzilla.libsdl.org/show_bug.cgi?id=5337)).

If that's the case, then there is no way to solve this without changing those libraries so that they do "something else" when they detect that they are run in a Flatpak container. One possible answer for the "something else" is to talk to a portal, as requested in this issue. Another possible answer is for them to monitor `/dev` and `/sys` directly, like SDL 2 now does.

Recent versions of SDL 2 do not suffer from this, because they detect this situation automatically, and monitor `/dev` and `/sys` directly ([#5337](https://bugzilla.libsdl.org/show_bug.cgi?id=5337), [#5360](https://bugzilla.libsdl.org/show_bug.cgi?id=5360), [#5361](https://bugzilla.libsdl.org/show_bug.cgi?id=5361). Steam has a statically linked copy of a recent SDL 2, so it will have picked up the same fixes.

Proton 5.13+ does not suffer from this either, because some of its code paths use SDL 2, and the others have had similar changes.

Native Linux games on Steam often bundle their own copy of SDL 2, so they will not necessarily have those changes. (This is unfortunate, and they should really use the copy in the Steam Runtime, which is suitably recent; but that's up to those games' developers and publishers.)

I don't know the status of Firefox or Chromium.

**Comment by HugLifeTiZ** [comment]
Thanks for all that info, it was really helpful. More than you know, in fact...

> I think this is most likely because SDL1 and those other means of reading the game controllers are relying on `udev_monitor` in `libudev`, which does not usually work in containers, because the events that it delivers do not cross the sandbox/container boundary ([more info here](https://bugzilla.libsdl.org/show_bug.cgi?id=5337)).

> If the container doesn't expose /run/udev, then obviously libudev isn't entirely going to work as intended.

:thinking:

```
sudo flatpak override --filesystem=/run/udev com.valvesoftware.Steam
```

And suddenly all the games I mentioned in my initial issue report have functioning gamepad input. When I apply this fix, Firefox does too. I have to rebuild OBS with browser support, but I wouldn't be surprised if it works there, too.

I am sure your response to that revelation is going to be "oh my god **please don't do that**," which is reasonable considering what you're trying to accomplish, but uh... as an end user, this "fixes" my problem.

**Comment by smcv** [comment]
> I am sure your response to that revelation is going to be "oh my god please don't do that"

Less extreme than that, perhaps, but...

You should be aware that this works for you by pure luck, and might not work tomorrow. The version of udev on the host does not claim to be compatible with the version of libudev in the container - apparently in your case, *today*, they're close enough. Tomorrow, or next week, or next year, who knows? But probably not.

This will also not be enough to make controller hotplugging work.

**Comment by smcv** [comment]
"Oh my god please don't do that" is tempting as a response to using SDL 1 this decade, though :-)

**Comment by HugLifeTiZ** [comment]
> You should be aware that this works for you by pure luck, and might not work tomorrow. The version of udev on the host does not claim to be compatible with the version of libudev in the container - apparently in your case, today, they're close enough. Tomorrow, or next week, or next year, who knows? But definitely not forever.

All of that is entirely fair, and when it does break in the future, I'll remember that.

> This will also not be enough to make controller hotplugging work.

Understandable. MoltenGamepad happens to handle that for me by means of having four virtual pads exist at all times.

> "Oh my god please don't do that" is tempting as a response to using SDL 1 this decade, though :-)

Trust me, I strongly agree, but tell that to every developer who used Unity 4, OpenTK, or some other old tech, and then never touched their Linux version again after shoving it out into the world. I mean, for all we know, they didn't use SDL *at all* and used their own code to handle what it does. :skull:

The "wrong way" works today, but the "right way" that we're working on for tomorrow: is one of its goals to support these legacy games somehow?

**Comment by hadess** [comment]
> And suddenly all the games I mentioned in my initial issue report have functioning gamepad input. When I apply the override to it, Firefox does too. I have to rebuild OBS with browser support, but I wouldn't be surprised if it works there, too. EDIT: No dice for Chromium nor for either version of Linux browser plugin on OBS.

It'd be easier to use `LD_PRELOAD` to cover those cracks. Or modify SDL1. (I've packaged a 32-bit closed source app for Flathub).

**Comment by HugLifeTiZ** [comment]
> It'd be easier to use `LD_PRELOAD` to cover those cracks. Or modify SDL1. (I've packaged a 32-bit closed source app for Flathub).

That makes sense. Does a library already exist that does gamepad enumeration "the right way", or is that yet to be created?

**Comment by hadess** [comment]
How could there be a library to access a portal that isn't implemented yet?

**Comment by HugLifeTiZ** [comment]
When you put it in terms of saying that "the right way" is the portal, that makes sense. But other applications can already enumerate gamepads without us poking a hole to expose /run/udev, and that's what I was thinking of as "the right way", though it is not as "right" as the portal would be. So do we presently have the ability to override the behavior of applications that are doing it the wrong way, to have them do it the "not wrong, but not quite right" way?

**Comment by smcv** [comment]
SDL 2 is likely to be the closest thing there is to doing gamepad enumeration the right way (disclaimer: I contributed many of the patches that made this happen, so obviously I'm biased).

libmanette is also reasonably close.

**Comment by smcv** [comment]
> do we presently have the ability to override the behavior of applications that are doing it the wrong way, to have them do it the "not wrong, but not quite right" way?

No. That would require either changing their source code, or injecting a `LD_PRELOAD` that reimplements the API they use (which can only work if they use a suitable API that can be reimplemented).

**Comment by HugLifeTiZ** [comment]
Right, and we definitely don't have the source code for the hundreds of legacy Linux games enumerating gamepads the wrong way. What about maybe creating a fake /run/udev? Is that what SrtInputMonitor's abstraction over udev does?

**Comment by HugLifeTiZ** [comment]
> If legacy games are using SDL1, then the least horrific route would be a modified SDL1 that does what SDL2 does, or a LD_PRELOAD that interposes replacements for SDL1's game controller enumeration functions that do what SDL2 does.

I *hope* that most Unity 4.x games actually are using SDL1 and aren't just imitating what it does... Ack, and then you have the games that use plugins for "better gamepad support" and then we have to wonder if *those* are using SDL1.

I thought I had heard about a project for implementing SDL1 on top of SDL2. Is that still a thing?

> If legacy games aren't using SDL1, I don't see any solution other than applying individual LD_PRELOAD hacks to each game (if that's even feasible).

I think at that point it might be better to run the Windows versions of those games via Proton.

**Comment by smcv** [comment]
> I thought I had heard about a project for implementing SDL1 on top of SDL2. Is that still a thing?

https://github.com/libsdl-org/sdl12-compat

> I think at that point it might be better to run the Windows versions of those games via Proton. That's what I have to do for CrossCode, even on regular distro Steam, since nw.js's Linux gamepad code is just so busted.

There's a finite number of past games and an unbounded number of future games, and we can influence how the future games work, so a solution that only works for the future games is a lot better than nothing.

**Comment by HugLifeTiZ** [comment]
> https://github.com/libsdl-org/sdl12-compat

Awesome, and still under active development. That will certainly prove useful to this cause.

> There's a finite number of past games and an unbounded number of future games, and we can influence how the future games work, so a solution that only works for the future games is a lot better than nothing.

I do ultimately agree with this, at the end of the day. I wouldn't want to hold back the future for the sake of the past. Preserving the past is part of the point of Proton anyways.

**Comment by rmader** [comment]
> This was rejected in favour of a Wayland protocol extension.
> 
> See "inputfd":
> https://gitlab.gnome.org/GNOME/mutter/-/tree/wip/carlosg/inputfd
> https://lists.freedesktop.org/archives/wayland-devel/2017-August/034828.html
> (which somebody needs to finish)

Here's is new MR for this: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/110

**Comment by guarapicci** [comment]
It's been 4 years. Just in case anyone's still on this, i have a suggestion.
- Keep a privileged "HID proxy" process polling any hidraw nodes.
- Portal asks explicit permission for input device access. Have 1 global control, or a list with n checkboxes.
- Every app gets its own hidraw node inside the container, for each controller it's been allowed access.
- Only feed HID events to the hidraw nodes of the currently focused application, or whatever application the user allowed "background input" for.
- For consistent layouts, let the user pick a uhid (virtual input device) remapped by moltenjoystick, 

I'm not familiar with flatpak, so i don't know if it's even possible to figure out which wayland/X11 window belongs to which process.

**Comment by swick** [comment]
Logind will soon allow access to hidraw devices which will allow compositors to implement the inputfd wayland protocol for evdev and hidraw devices. Just needs someone to implement it.

**Comment by whot** [comment]
Sorry to potentially side-line but has there been a consensus on whether it's going to be a device portal or inputfd? I've kinda lost track of the various bugs/rfcs.

**Comment by swick** [comment]
The wayland protocol ties everything to focus which means applications can get access without any permission dialog popping up etc. and that's why it's something we definitely want because it would allow us to make the DE and apps usable with gamepads and might prove useful for games but it probably won't satisfy all use cases right away.

**Comment by ssokolow** [comment]
> but it probably won't satisfy all use cases right away.

Agreed. Before I just said "f*** it!" and "fixed" things by hanging a whole separate computer off my central monitor via a KVM switch (both a hardware one and [Barrier](https://github.com/debauchee/barrier) plus the monitor's internal switcher, depending on whether I want to play a game with mouselook) to run the games (complemented by leaving my ATi Remote Wonder II un-switched to control the wing monitors), what I wanted was:

1. Desktop recognizes gamepad input as user activity so games don't need to take the brute-force approach of suppressing the screensaver as if they're media players. (I've never seen the screensaver kick in during a gamepad-controled game on the Windows 98SE, Windows XP, or Windows 7 machines in my hobby corner. Be at least that good.)
2. Game receives gamepad input when not focused so I can give keyboard focus to something like Discord on one of my wing monitors and leave the mouse over a GOG.com manual or strategy guide PDF on the other wing monitor. (One of the greastest things X11 ever did was normalize how every toolkit except GTK3 lets you send scroll-wheel events to windows without focusing them first. More generally, I'd like to see games not being able to know whether they've lost focus unless the desktop allows it, so I can globally veto auto-pause-on-loss-of-focus. That "split focus" behaviour is one of the things that gets fixed by either running the game in a nested desktop a la Xephyr or Wine virtual desktop or using a KVM switch. IIRC, the Mupen64plus emulator (yes, I do own a Retrode2 to dump my childhood carts now that my childhood N64 has a reset problem) found a new way to be frustrating by not pausing on loss of focus, and always listening to the axes on the gamepad... but requiring focus to listen to the buttons!)

I can certainly see the value of "gamepad focus works the same way as keyboard and mouse focus" though. Given that I don't want every game to have to manually support either case, probably best to implement a solution that's user-selectable.

**Comment by swick** [comment]
The focus for inputfd would be a different focus from mouse which is different from keyboard. We have thought about making it possible in GNOME to "stick" the inputfd focus to a specific window. I think that would satisfy most use cases and which that the remaining big question is backwards compatibility.

**Comment by ssokolow** [comment]
I don't use GNOME for unrelated reasons but that sounds like a design decision that would meet my needs if KDE copied it.

(To be honest, it sounds so natural to put the UI for that in KWin's SSD context menu and SSD titlebar button customizer KCM that I'm surprised to hear it being considered in GNOME in any form.)
